### PR TITLE
Run IBCMerge on Linux build

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
* Restores IBC data for both Linux and Windows on all platforms
* Chooses the right IBC data based on the target OS
* When target OS isn't anything specific (like, e.g. System.Linq and other assemblies that are not OS specific), embeds Windows IBC data

### CODE REVIEW

**Comment by MichalStrehovsky** [comment]
Cc @sergiy-k @RussKeldorph @brianrob

File: eng/codeOptimization.targets [hunk 1]
```diff
@@ -9,6 +9,10 @@
   <Target Name="SetApplyNgenOptimization"
           Condition="'$(IsEligibleForNgenOptimization)' == 'true'"
           BeforeTargets="CoreCompile">
+    <PropertyGroup>
+      <IbcOptimizationDataDir Condition="'$(OSGroup)' == 'Unix'">$(IbcOptimizationDataDir)$(LinuxCoreFxOptimizationDataPackageId)\</IbcOptimizationDataDir>
```
Comment (ericstj) [inline eng/codeOptimization.targets]: You may want to consider more than just Unix. https://github.com/dotnet/corefx/blob/master/eng/configurations/osgroups.props

**Comment by Wraith2** [comment]
What is IBC data?

**Comment by filipnavara** [comment]
> What is IBC data?

Citing @MichalStrehovsky from Gitter: "IBC = Instrumented Block Counts. Basically, profile guided optimizations for managed code."

**Comment by ViktorHofer** [comment]
This broke the official builds: https://dnceng.visualstudio.com/internal/_build/results?buildId=151325

**Comment by MichalStrehovsky** [comment]
Well that's interesting.

```
/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/1.0.0-beta.19207.1/tools/OptimizationData.targets(176,5): error MSB3073: The command ""/__w/1/s/.dotnet/dotnet" exec "/__w/1/s/.packages/microsoft.dotnet.ibcmerge/5.0.6-beta.19203.1/tools/netcoreapp2.0/ibcmerge.dll" -q -f  -minify -delete -mo "/__w/1/s/artifacts/tmp/netcoreapp-Release/OptimizedAssemblies/f5846cd9-09cb-4bff-80f4-5a9cc2e9169e" "/__w/1/s/artifacts/ibc/optimization.windows_nt-x64.IBC.CoreFx/data/System.Collections.NonGeneric.dll/System.Collections.NonGeneric.ibc"" exited with code 129. [/__w/1/s/src/System.Collections.NonGeneric/src/System.Collections.NonGeneric.csproj]
  The application to execute does not exist: '/__w/1/s/.packages/microsoft.dotnet.ibcmerge/5.0.6-beta.19203.1/tools/netcoreapp2.0/ibcmerge.dll'
```

The way this is supposed to work (and I actually tested this on both Windows and Linux in both configs) is that there are two options:
1. Ibcmerge.dll was not restored (because we didn't restore the internal tools). We don't run it and just print a message with the arguments we would use to invoke it.
2. Ibcmerge.dll exists. We run it.

Obviously, there's a secret option 3: Ibcmerge was restored on the official build machine where ibcmerge.dll has a quantum state. This does not spark joy.

**Comment by MichalStrehovsky** [comment]
Rollback in #36791.

**Comment by ericstj** [comment]
Just had a look at the different steps and there's definitely something wrong here:

https://dev.azure.com/dnceng/internal/_build/results?buildId=151325&view=logs&jobId=7cc86366-6ece-5781-60ca-3d1c814b1449&taskId=7388ea57-952b-5299-49bb-4b8b88a97717&lineStart=8&lineEnd=16&colStart=1&colEnd=48

https://dev.azure.com/dnceng/internal/_build/results?buildId=151325&view=logs&jobId=338934ee-6440-5778-bafb-9371fbcccfa1&taskId=a2a8a210-de79-5606-58f6-9ddff1384723&lineStart=15&lineEnd=15&colStart=1&colEnd=260

So it looks like the Linux legs don't even attempt to run nuget at all.

**Comment by ViktorHofer** [comment]
Yes internal tools restore currently only works on Windows as we use nuget.exe in the yml step. Locally we use a crossplat solution. I will check what we can do here.

**Comment by safern** [comment]
> use a crossplat solution. I will check what we can do here.

I'm already in it... @chcosta is helping me set some service connections or an `upstream` in dnceng that uses the feed. Instead of using nuget.exe, I have changed it locally to use the .net core task to restore.

**Comment by sergiy-k** [comment]
Given that https://github.com/dotnet/corefx/pull/36840 has been merged, can we try enabling IBCMerge on Linux again?

**Comment by ViktorHofer** [comment]
Yes 👍

**Comment by MichalStrehovsky** [comment]
#36924 has the revert of the revert.

**Pull request merged**